### PR TITLE
Try to checkout matching gzdev branch for PR jobs

### PR DESCRIPTION
**Pull request by scpeters**
If a GitHub pull request branch contains `matching_branch/`, then try to checkout that branch from `gzdev`. To simplify checking out other branches, clone gzdev without `--depth 1`.

Part of #564.

Testing:

* https://github.com/ignitionrobotics/ign-fuel-tools/pull/218
* https://github.com/ignition-tooling/gzdev/compare/scpeters/matching_branch/test_msgs9?expand=1
* [Build Status](https://build.osrfoundation.org/job/ignition_fuel-tools-ci-pr_any-ubuntu_auto-amd64/673/) https://build.osrfoundation.org/job/ignition_fuel-tools-ci-pr_any-ubuntu_auto-amd64/673/

### CODE REVIEW

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash [hunk 1]
```diff
@@ -33,24 +33,32 @@ export APT_PARAMS=
...
+  GZDEV_TRY_BRANCH=$ghprbSourceBranch
+fi
 
 dockerfile_install_gzdev_repos()
 {
 cat >> Dockerfile << DELIM_OSRF_REPO_GIT
 RUN rm -fr ${GZDEV_DIR}
-RUN git clone --depth 1 https://github.com/ignition-tooling/gzdev -b ${GZDEV_BRANCH} ${GZDEV_DIR}
+RUN git clone https://github.com/ignition-tooling/gzdev -b ${GZDEV_BRANCH} ${GZDEV_DIR}
+RUN bash -c "if [[ -n \"$GZDEV_TRY_BRANCH\" ]]; then git -C ${GZDEV_DIR} fetch origin \"$GZDEV_TRY_BRANCH\"; fi || true"
```
Comment (j-rivero) [inline jenkins-scripts/docker/lib/docker_generate_dockerfile.bash:45]: This is starting to be a bit complicate to read. Do we need to use bash here or can we implement that using plan sh, something like:
```bash
RUN if [ -n foo  ]; then \
      false; \
    fi || true
```

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash [hunk 2]
```diff
@@ -33,24 +33,32 @@ export APT_PARAMS=
...
+RUN bash -c "if [[ -n \"$GZDEV_TRY_BRANCH\" ]]; then git -C ${GZDEV_DIR} checkout \"$GZDEV_TRY_BRANCH\"; fi || true"
```
Comment (j-rivero) [inline jenkins-scripts/docker/lib/docker_generate_dockerfile.bash:46]: Could we merge this git command with the previous to save the duplication of the if?

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash [hunk 3]
```diff
@@ -33,24 +33,32 @@ export APT_PARAMS=
 
 GZDEV_DIR=${WORKSPACE}/gzdev
 GZDEV_BRANCH=${GZDEV_BRANCH:-master}
+if [[ "$ghprbSourceBranch" =~ matching_branch\/ ]]; then
```
Comment (j-rivero) [inline jenkins-scripts/docker/lib/docker_generate_dockerfile.bash:36]: Using somehow the branch name to include some internal commands is not a bad idea. Could we somehow introduce the notion of "this is for Jenkins" or "this is a command for infra" in the name? Something like branch name: `*/ci_matching_branch/*` or similar?

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash [hunk 4]
```diff
@@ -33,24 +33,32 @@ export APT_PARAMS=
...
+RUN bash -c "if [[ -n \"$GZDEV_TRY_BRANCH\" ]]; then git -C ${GZDEV_DIR} fetch origin \"$GZDEV_TRY_BRANCH\"; fi || true"
+RUN bash -c "if [[ -n \"$GZDEV_TRY_BRANCH\" ]]; then git -C ${GZDEV_DIR} checkout \"$GZDEV_TRY_BRANCH\"; fi || true"
+RUN git -C ${GZDEV_DIR} branch
 DELIM_OSRF_REPO_GIT
 if [[ -n ${GZDEV_PROJECT_NAME} ]]; then
 # debian sid docker images does not return correct name so we need to use
 # force-linux-distro
 cat >> Dockerfile << DELIM_OSRF_REPO_GZDEV
 RUN ${GZDEV_DIR}/gzdev.py repository enable --project=${GZDEV_PROJECT_NAME} --force-linux-distro=${DISTRO} || ( git -C ${GZDEV_DIR} pull origin ${GZDEV_BRANCH} && \
+    (bash -c "if [[ -n \"$GZDEV_TRY_BRANCH\" ]]; then git -C ${GZDEV_DIR} checkout \"$GZDEV_TRY_BRANCH\"; fi || true") && \
```
Comment (j-rivero) [inline jenkins-scripts/docker/lib/docker_generate_dockerfile.bash:54]: The whole RUN command is really hard to read. We need to find a better way.

**Review by j-rivero — CHANGES_REQUESTED** [review]

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash (hunk 3 above)
Comment (scpeters) [reply jenkins-scripts/docker/lib/docker_generate_dockerfile.bash:36]: sure; I'll go for `ci_matching_branch/`, so it doesn't have to have a preceding `/`

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash [hunk 5]
```diff
@@ -33,24 +33,32 @@ export APT_PARAMS=
...
 
 dockerfile_install_gzdev_repos()
 {
 cat >> Dockerfile << DELIM_OSRF_REPO_GIT
 RUN rm -fr ${GZDEV_DIR}
-RUN git clone --depth 1 https://github.com/ignition-tooling/gzdev -b ${GZDEV_BRANCH} ${GZDEV_DIR}
+RUN git clone https://github.com/ignition-tooling/gzdev -b ${GZDEV_BRANCH} ${GZDEV_DIR}
+RUN bash -c "if [[ -n \"$GZDEV_TRY_BRANCH\" ]]; then git -C ${GZDEV_DIR} fetch origin \"$GZDEV_TRY_BRANCH\"; fi || true"
+RUN bash -c "if [[ -n \"$GZDEV_TRY_BRANCH\" ]]; then git -C ${GZDEV_DIR} checkout \"$GZDEV_TRY_BRANCH\"; fi || true"
+RUN git -C ${GZDEV_DIR} branch
```
[the remark this answers is not in the document]
Comment (scpeters) [reply jenkins-scripts/docker/lib/docker_generate_dockerfile.bash:47]: this is just a debugging print statement

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash (hunk 1 above)
Comment (scpeters) [reply jenkins-scripts/docker/lib/docker_generate_dockerfile.bash]: I think at one point I had the `"$ghprbSourceBranch" =~ ci_matching_branch\/` logic in every `RUN` statement, which requires bash, but then I moved it outside the `RUN` and forgot to revert to `sh` syntax. Thanks for noticing that; I've tried to simplify it in https://github.com/ignition-tooling/release-tools/pull/572/commits/2c6faf12af27a5caf69e208d9eceb3dfe1382e5b; let me know if it still needs more work

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash (hunk 2 above)
Comment (scpeters) [reply jenkins-scripts/docker/lib/docker_generate_dockerfile.bash]: merged in https://github.com/ignition-tooling/release-tools/pull/572/commits/2c6faf12af27a5caf69e208d9eceb3dfe1382e5b

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash (hunk 3 above)
Comment (scpeters) [reply jenkins-scripts/docker/lib/docker_generate_dockerfile.bash]: done in https://github.com/ignition-tooling/release-tools/pull/572/commits/2c6faf12af27a5caf69e208d9eceb3dfe1382e5b

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash (hunk 4 above)
Comment (scpeters) [reply jenkins-scripts/docker/lib/docker_generate_dockerfile.bash]: is it better after https://github.com/ignition-tooling/release-tools/pull/572/commits/2c6faf12af27a5caf69e208d9eceb3dfe1382e5b?

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash (hunk 5 above)
[the remark this answers is not in the document]
Comment (j-rivero) [reply jenkins-scripts/docker/lib/docker_generate_dockerfile.bash:49]: oook!
```suggestion
# print branch for information proposes 
RUN git -C ${GZDEV_DIR} branch
```

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash (hunk 1 above)
Comment (j-rivero) [reply jenkins-scripts/docker/lib/docker_generate_dockerfile.bash]: wow! much better thanks!

**Review by j-rivero — APPROVED** [review]
Minor suggestion for a comment but the final result is great. Thanks Steve!

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash (hunk 5 above)
[the remark this answers is not in the document]
Comment (scpeters) [reply jenkins-scripts/docker/lib/docker_generate_dockerfile.bash:50]: https://github.com/ignition-tooling/release-tools/pull/572/commits/b03ecc2ab70905094db06c1660123aac4ce09c0c

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash [hunk 6]
```diff
@@ -33,24 +33,35 @@ export APT_PARAMS=
...
 dockerfile_install_gzdev_repos()
 {
 cat >> Dockerfile << DELIM_OSRF_REPO_GIT
 RUN rm -fr ${GZDEV_DIR}
-RUN git clone --depth 1 https://github.com/ignition-tooling/gzdev -b ${GZDEV_BRANCH} ${GZDEV_DIR}
+RUN git clone https://github.com/ignition-tooling/gzdev -b ${GZDEV_BRANCH} ${GZDEV_DIR}
+RUN if [ -n $GZDEV_TRY_BRANCH ]; then \
+        git -C ${GZDEV_DIR} fetch origin $GZDEV_TRY_BRANCH || true; \
+        git -C ${GZDEV_DIR} checkout $GZDEV_TRY_BRANCH || true; \
+    fi || true"
```
Comment (nuclearsandwich) [inline jenkins-scripts/docker/lib/docker_generate_dockerfile.bash:48]: The stray quote here was removed via direct push https://github.com/ignition-tooling/release-tools/commit/99524e1347c272920f4b384f3372037d3d38bf36

**Pull request merged**